logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/slow/SlowSyncManager.kt

```diff
@@ -50,35 +54,61 @@ internal class SlowSyncManager(
         }
     })
 
-    init { startMonitoring() }
+    init {
+        startMonitoring()
+    }
 
     private fun startMonitoring() {
         scope.launch(coroutineExceptionHandler) {
             slowSyncCriteriaProvider
                 .syncCriteriaFlow()
                 .distinctUntilChanged()
+                .combine(slowSyncRepository.observeLastSlowSyncCompletionInstant())
                 // Collect latest will cancel whatever is running inside the collector when a new value is emitted
-                .collectLatest { handleCriteriaResolution(it) }
+                .collectLatest { (syncCriteriaResolution, lastTimeSlowSyncWasPerformed) ->
+                    handleCriteriaResolution(syncCriteriaResolution, lastTimeSlowSyncWasPerformed)
+                }
         }
     }
 
-    private suspend fun handleCriteriaResolution(it: SyncCriteriaResolution) {
-        if (it is SyncCriteriaResolution.Ready) {
-            // START SYNC
-            logger.i("Starting SlowSync as all criteria are met")
-            slowSyncWorker.performSlowSyncSteps().cancellable().collect { step ->
-                logger.i("Performing SlowSyncStep $step")
-                slowSyncRepository.updateSlowSyncStatus(SlowSyncStatus.Ongoing(step))
+    private suspend fun handleCriteriaResolution(syncCriteriaResolution: SyncCriteriaResolution, lastTimeSlowSyncWasPerformed: Instant?) {
+        if (syncCriteriaResolution is SyncCriteriaResolution.Ready) {
+            // START SYNC IF NEEDED
+            logger.i("SlowSync criteria ready, checking if SlowSync is needed or already performed")
+            logger.i("Last SlowSync was performed on '$lastTimeSlowSyncWasPerformed'")
+            if (isSlowSyncNeeded(lastTimeSlowSyncWasPerformed)) {
+                logger.i("Starting SlowSync as all criteria are met and it wasn't performed recently")
+                performSlowSync()
+            } else {
+                logger.i("No need to perform SlowSync. Marking as Complete")
             }
             slowSyncRepository.updateSlowSyncStatus(SlowSyncStatus.Complete)
+            slowSyncRepository.setLastSlowSyncCompletionInstant(Clock.System.now())
```

Comment:
Shouldn't this only be updated inside the `if (isSlowSyncNeeded(lastTimeSlowSyncWasPerformed))`?

Reply:
You are absolutely correct and this was a major oversight when I was moving code around.
Nice catch.

I'll add a test to check for this as well.

Reply:
Done